### PR TITLE
 JDK-8301225: Replace NULL with nullptr in share/gc/shenandoah/

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory share/gc/shenandoah/. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301225](https://bugs.openjdk.org/browse/JDK-8301225): Replace NULL with nullptr in share/gc/shenandoah/

### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - no project role) ⚠️ Review applies to [8e2596a9](https://git.openjdk.org/jdk/pull/12251/files/8e2596a9f241558ba40c87dc81a9f724b042dc05)
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role) ⚠️ Review applies to [8e2596a9](https://git.openjdk.org/jdk/pull/12251/files/8e2596a9f241558ba40c87dc81a9f724b042dc05)
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12251/head:pull/12251` \
`$ git checkout pull/12251`

Update a local copy of the PR: \
`$ git checkout pull/12251` \
`$ git pull https://git.openjdk.org/jdk pull/12251/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12251`

View PR using the GUI difftool: \
`$ git pr show -t 12251`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12251.diff">https://git.openjdk.org/jdk/pull/12251.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp [hunk 1]
```diff
@@ -256,7 +256,7 @@ void ShenandoahBarrierSetC2::satb_write_barrier_pre(GraphKit* kit,
       pre_val = __ load(__ ctrl(), adr, val_type, bt, alias_idx);
     }
 
-    // if (pre_val != NULL)
+    // if (pre_val != null)
```
Comment (jdksjolen) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:259]: nullptr

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp [hunk 2]
```diff
@@ -278,13 +278,13 @@ void ShenandoahBarrierSetC2::satb_write_barrier_pre(GraphKit* kit,
         const TypeFunc *tf = ShenandoahBarrierSetC2::write_ref_field_pre_entry_Type();
         __ make_leaf_call(tf, CAST_FROM_FN_PTR(address, ShenandoahRuntime::write_ref_field_pre_entry), "shenandoah_wb_pre", pre_val, tls);
       } __ end_if();  // (!index)
-    } __ end_if();  // (pre_val != NULL)
+    } __ end_if();  // (pre_val != null)
```
Comment (jdksjolen) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:281]: nullptr

File: src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp [hunk 3]
```diff
@@ -425,15 +425,15 @@ void ShenandoahBarrierSetC2::insert_pre_barrier(GraphKit* kit, Node* base_oop, N
       __ sync_kit(kit);
 
       Node* one = __ ConI(1);
-      // is_instof == 0 if base_oop == NULL
+      // is_instof == 0 if base_oop == null
```
Comment (jdksjolen) [inline src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:428]: nullptr

File: src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.cpp [hunk 4]
```diff
@@ -55,10 +55,10 @@ ShenandoahPhaseTimings::ShenandoahPhaseTimings(uint max_workers) :
...
   // Initialize everything to sane defaults
   for (uint i = 0; i < _num_phases; i++) {
-#define SHENANDOAH_WORKER_DATA_NULL(type, title) \
-    _worker_data[i] = NULL;
-    SHENANDOAH_PAR_PHASE_DO(,, SHENANDOAH_WORKER_DATA_NULL)
-#undef SHENANDOAH_WORKER_DATA_NULL
+#define SHENANDOAH_WORKER_DATA_nullptr(type, title) \
+    _worker_data[i] = nullptr;
+    SHENANDOAH_PAR_PHASE_DO(,, SHENANDOAH_WORKER_DATA_nullptr)
+#undef SHENANDOAH_WORKER_DATA_nullptr
```
Comment (jdksjolen) [inline src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.cpp:61]: Fix these macros

**Review by jdksjolen — COMMENTED** [review]
Manual stuff to fix

**Comment by jdksjolen** [comment]
Passes tier1.

**Review by earthling-amzn — APPROVED** [review]
LGTM

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:2]: minor glitch here

File: src/hotspot/share/gc/shenandoah/shenandoahMarkBitMap.hpp [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahMarkBitMap.hpp:2]: Probably should be 2018, 2019, 2023

File: src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015, 2021, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (kdnilsen) [inline src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.cpp:2]: same here

**Review by kdnilsen — COMMENTED** [review]
Looks good to me.  Thanks for doing this.

A few comments about copyright notices.

**Review by kdnilsen — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.cpp (hunk 4 above)
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.cpp]: My suggestion would be to leave the macro name unchanged, just change the definition. Or at least use uppercasing in the macro name. (Perhaps the latter's what your comment intended.)

More generally, is there a `jcheck` rule that will prevent re-introduction of NULL usage into the mix?

Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahPhaseTimings.cpp]: This code diff is out of date, the macro name was reverted to the original in a subsequent commit. Good question about adding this to `jcheck`.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp (hunk 5 above)
Comment (jdksjolen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:2]: Weird, first time I'm seeing this glitch. Thanks!

File: src/hotspot/share/gc/shenandoah/shenandoahMarkBitMap.hpp (hunk 6 above)
Comment (jdksjolen) [reply src/hotspot/share/gc/shenandoah/shenandoahMarkBitMap.hpp:2]: This and the other ones are correct, it's `$FirstChange, $LastChange, `

**Comment by jdksjolen** [comment]
Trivial merge conflict fix and copyright fix. Tier1 passes, integrating.

/integrate

**Review by rkennke — APPROVED** [review]
Looks good to me, thank you!

**Comment by jdksjolen** [comment]
There we go, let's integrate!

/integrate

**Pull request closed without merging**